luacompiler/src/lib/frontend/mod.rs

```diff
@@ -0,0 +1,174 @@
+pub mod constants_map;
+pub mod register_map;
+use self::{constants_map::ConstantsMap, register_map::RegisterMap};
+use bytecode::{
+    instructions::{make_instr, Opcode},
+    LuaBytecode,
+};
+use cfgrammar::RIdx;
+use lrpar::Node::{self, *};
+use lua5_3_l;
+use lua5_3_y;
+use LuaParseTree;
+
+/// Represents a compiler which translates a given Lua parse tree to some bytecode representation.
+/// This compiler will be changed in the future to translate from lua to a higher-level
+/// representation, which is easier to translate to the current bytecode.
+pub struct LuaToBytecode<'a> {
+    pt: &'a LuaParseTree,
+}
+
+impl<'a> LuaToBytecode<'a> {
+    pub fn new(pt: &'a LuaParseTree) -> LuaToBytecode {
+        LuaToBytecode { pt }
+    }
+
+    /// Compile the parse tree to an intermediate representation.
+    pub fn compile_to_ir(&self) -> LuaBytecode {
+        let mut instrs = vec![];
+        let mut pt_nodes: Vec<&Node<u8>> = vec![&self.pt.tree];
+        let mut reg_map = RegisterMap::new();
+        let mut const_map = ConstantsMap::new();
+        while !pt_nodes.is_empty() {
+            let node = pt_nodes.pop().unwrap(); // always checked if it is empty
+            match *node {
+                Nonterm {
+                    ridx: RIdx(ridx),
+                    ref nodes,
+                } if ridx == lua5_3_y::R_STAT => {
+                    debug_assert!(nodes.len() == 3);
+                    match nodes[1] {
```

Comment:
This feels like it would be more concise as `if let`.

Reply:
the reason I am doing a match is so that it is easier to extract nodes in the case of `Nonterm`. I feel like having `if let ... else if let ...` is harder to read than it is currently.

Reply:
Yes but your `else` branch here is effectively empty `_ => {}`. That said, I can't remember if `if let` can have a side `if` condition; if not, perhaps `match` is best.

Reply:
Unfortunately you can't chain things in an `if let` statement with `&&`. The RFC for this was approved, but the compiler outputs: `note: this will be a error until the let_chains feature is stabilized`

Reply:
Fair enough!